### PR TITLE
feat: Kernel fix & test improvements/refactoring

### PR DESCRIPTION
**Pull request by dbanks12**
# Description

* Blocks https://github.com/AztecProtocol/aztec3-packages/pull/263
* Closes https://github.com/AztecProtocol/aztec3-circuits/issues/172
* Closes https://github.com/AztecProtocol/aztec3-circuits/issues/213
* Some work related to: https://github.com/AztecProtocol/aztec3-circuits/issues/173
* Fixes bug introduced in: https://github.com/AztecProtocol/aztec3-circuits/pull/167
  * contract logic wasn't being called at all in native kernel...
  * also makes tests more robust to catch such bugs (tests input real contract tree root and check contract address output from circuit)

## Status
* ~Blocked by https://github.com/AztecProtocol/aztec3-circuits/pull/233~
* ~Needs to be rebased on master which will be some work.~
* I want to decide if there is a better way to handle private circuit call args than the way I am in this PR.
  * Decided to punt this for now since this issue now blocks https://github.com/AztecProtocol/aztec3-packages/pull/263
  * Created issue for this: https://github.com/AztecProtocol/aztec3-circuits/issues/246
* ~Need to comment helper functions in hash.hpp.~

## Contents
* Kernel tests have been refactored to use helper functions.
* Hash/tree functions have been moved into helpers in hash.hpp (for kernel at least).
* Fixed native kernel circuit which was not calling any contract logic.
* Changed the way args are passed to private circuit calls so that we can use a `std::function` type to write functions that accept any circuit function as input.

# Checklist:

- [ ] I have reviewed my diff in github, line by line.
- [ ] Every change is related to the PR description.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this pull request to the issue(s) that it resolves.
- [ ] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [ ] There are no circuit changes, OR specifications in `/specs` have been updated.
- [ ] There are no circuit changes, OR a cryptographer has been assigned for review.
- [ ] I've updated any terraform that needs updating (e.g. environment variables) for deployment.
- [ ] The branch has been rebased against the head of its merge target.
- [ ] I'm happy for the PR to be merged at the reviewer's next convenience.
- [ ] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [ ] If existing code has been modified, such documentation has been added or updated.

> **Note**
> If you are updating the submodule, please make sure you do it in its own _special_ PR and avoid making changes to the submodule as a part of other PRs.
> To update a submodule, you can run the following commands:
> ```console
> $ git submodule update --recursive
> ```
> Alternatively, you can select a particular commit in `barretenberg/aztec3` that you wish to point to:
> ```console
> $ cd barretenberg
> $ git pull origin aztec3        # This will point to the latest commit in `barretenberg/aztec3`
> $ git checkout <commit_hash>    # Use this if you wish to point to a particular commit.
> $ cd ..
> $ git add . && git commit -m <commit_msg>
> $ git push
> ```

### CODE REVIEW

File: cpp/src/aztec3/circuits/kernel/private/.test.cpp [hunk 1]
```diff
@@ -70,288 +61,256 @@ using aztec3::circuits::apps::test_apps::escrow::deposit;
 
 using DummyComposer = aztec3::utils::DummyComposer;
 
-// using aztec3::circuits::mock::mock_circuit;
 using aztec3::circuits::mock::mock_kernel_circuit;
 
+// A type representing any private circuit function
+// (for now it works for deposit and constructor)
+using private_function = std::function<OptionalPrivateCircuitPublicInputs<NT>(
```
Comment (iAmMichaelConnor) [inline cpp/src/aztec3/circuits/kernel/private/.test.cpp:68]: I was going to ask why the args to the app circuits' function signatures were changed (in other files), and I guess this abstraction is the reason?

File: cpp/src/aztec3/circuits/kernel/private/.test.cpp [hunk 2]
```diff
@@ -70,288 +61,256 @@ using aztec3::circuits::apps::test_apps::escrow::deposit;
...
+#ifdef DEBUG_PRINTS
+    info("computed witness: ", composer.computed_witness);
+    // info("witness: ", private_kernel_composer.witness);
+    // info("constant variables: ", private_kernel_composer.constant_variables);
+    // info("variables: ", composer.variables);
 
-    Composer deposit_composer = Composer("../barretenberg/cpp/srs_db/ignition");
-    DB db;
+    // TODO: this fails intermittently, with:
+    // bigfield multiply range check failed
```
Comment (iAmMichaelConnor) [inline cpp/src/aztec3/circuits/kernel/private/.test.cpp:97]: I think this TODO has been fixed, so this comment can be removed.

File: cpp/src/aztec3/circuits/kernel/private/.test.cpp [hunk 3]
```diff
@@ -70,288 +61,256 @@ using aztec3::circuits::apps::test_apps::escrow::deposit;
...
-
-    const NT::fr msg_sender_private_key = 123456789;
-    const NT::address msg_sender =
-        NT::fr(uint256_t(0x01071e9a23e0f7edULL, 0x5d77b35d1830fa3eULL, 0xc6ba3660bb1f0c0bULL, 0x2ef9f7f09867fd6eULL));
-    const NT::address tx_origin = msg_sender;
+#ifdef DEBUG_PRINTS
+    info("computed witness: ", composer.computed_witness);
+    // info("witness: ", private_kernel_composer.witness);
+    // info("constant variables: ", private_kernel_composer.constant_variables);
+    // info("variables: ", composer.variables);
```
Comment (iAmMichaelConnor) [inline cpp/src/aztec3/circuits/kernel/private/.test.cpp:94]: Not sure we'll want to print these things anymore. I think I might have left these in here by accident when I was learning about the composer.

File: cpp/src/aztec3/circuits/kernel/private/.test.cpp [hunk 4]
```diff
@@ -70,288 +61,256 @@ using aztec3::circuits::apps::test_apps::escrow::deposit;
...
+ * @brief Generate a verification key for a private circuit.
+ *
+ * @details Use some dummy inputs just to get the VK for a private circuit
+ *
+ * @param is_constructor Whether this private call is a constructor call
+ * @param func The private circuit call to generate a VK for
+ * @param num_args Number of args to that private circuit call
+ * @return std::shared_ptr<NT::VK> - the generated VK
+ */
+std::shared_ptr<NT::VK> gen_func_vk(bool is_constructor, private_function const& func, size_t const num_args)
```
Comment (iAmMichaelConnor) [inline cpp/src/aztec3/circuits/kernel/private/.test.cpp:116]: We might be able to remove this function when we complete:
- https://github.com/AztecProtocol/aztec3-circuits/issues/209
and/or
- https://github.com/AztecProtocol/aztec3-circuits/issues/200

File: cpp/src/aztec3/circuits/kernel/private/.test.cpp (hunk 1 above)
Comment (dbanks12) [reply cpp/src/aztec3/circuits/kernel/private/.test.cpp:68]: That's right :thinking: trying to think of a better way. Do you think I overdid the generalization of this test helper logic?

File: cpp/src/aztec3/circuits/kernel/private/.test.cpp [hunk 5]
```diff
@@ -70,288 +61,256 @@ using aztec3::circuits::apps::test_apps::escrow::deposit;
...
-
-    const NT::address escrow_contract_address = 12345;
-    // const NT::fr escrow_contract_leaf_index = 1;
-    const NT::fr escrow_portal_contract_address = 23456;
-
-    const NT::fr msg_sender_private_key = 123456789;
-    const NT::address msg_sender =
-        NT::fr(uint256_t(0x01071e9a23e0f7edULL, 0x5d77b35d1830fa3eULL, 0xc6ba3660bb1f0c0bULL, 0x2ef9f7f09867fd6eULL));
-    const NT::address tx_origin = msg_sender;
+#ifdef DEBUG_PRINTS
```
[the remark this answers is not in the document]
Comment (dbanks12) [reply cpp/src/aztec3/circuits/kernel/private/.test.cpp:90]: It's a new directive that i created.... didn't know if I should use the same DEBUG flag that is used for other things (not sure what exactly).

File: cpp/src/aztec3/circuits/kernel/private/.test.cpp (hunk 4 above)
Comment (dbanks12) [reply cpp/src/aztec3/circuits/kernel/private/.test.cpp:116]: We need it for ContractDeploymentData and for function tree leaves

Comment (dbanks12) [reply cpp/src/aztec3/circuits/kernel/private/.test.cpp:116]: This generates the VK for a private call, not a mock kernel VK.

File: cpp/src/aztec3/circuits/kernel/private/.test.cpp [hunk 6]
```diff
@@ -409,20 +364,15 @@ TEST(private_kernel_tests, test_native_deposit)
 
     Prover mock_kernel_prover = mock_kernel_composer.create_prover();
     NT::Proof mock_kernel_proof = mock_kernel_prover.construct_proof();
+    // info("\nmock_kernel_proof: ", mock_kernel_proof.proof_data);
 
     std::shared_ptr<NT::VK> mock_kernel_vk = mock_kernel_composer.compute_verification_key();
```
Comment (dbanks12) [inline cpp/src/aztec3/circuits/kernel/private/.test.cpp:369]: @iAmMichaelConnor this is the big chunk we should be able to remove once we have the special kernel version for the initial iteration. Won't need to compute the mock kernel anymore.

File: cpp/src/aztec3/circuits/hash.hpp [hunk 7]
```diff
@@ -87,20 +92,108 @@ typename NCT::fr add_contract_address_to_nullifier(typename NCT::address contrac
  * @param siblingPath The nodes representing the merkle siblings of the leaf, its parent,
  * the next parent, etc up to the sibling below the root
  * @return The computed Merkle tree root.
+ *
+ * TODO need to use conditional assigns instead of `ifs` for circuit version.
+ *      see membership.hpp:check_subtree_membership (left/right/conditional_assign, etc)
```
[the remark this answers is not in the document]
Comment (iAmMichaelConnor) [reply cpp/src/aztec3/circuits/hash.hpp:97]: Just opened an issue for it.

File: cpp/src/aztec3/circuits/kernel/private/.test.cpp (hunk 4 above)
Comment (iAmMichaelConnor) [reply cpp/src/aztec3/circuits/kernel/private/.test.cpp:116]: Ah, thanks for explaining

**Pull request merged**